src/main/java/com/kobylynskyi/graphql/codegen/kotlin/KotlinGraphQLTypeMapper.java

```diff
@@ -0,0 +1,145 @@
+import com.kobylynskyi.graphql.codegen.model.MappingContext;
+import com.kobylynskyi.graphql.codegen.model.NamedDefinition;
+import com.kobylynskyi.graphql.codegen.model.graphql.GraphQLOperation;
+import com.kobylynskyi.graphql.codegen.utils.Utils;
+import graphql.language.Argument;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.kobylynskyi.graphql.codegen.java.JavaGraphQLTypeMapper.JAVA_UTIL_LIST;
+import static java.util.Arrays.asList;
+
+/**
+ * @author 梦境迷离
+ * @since 2020/12/09
+ */
+public class KotlinGraphQLTypeMapper implements GraphQLTypeMapper {
+
+    private static final String KOTLIN_UTIL_LIST = "List";
+    private static final String KOTLIN_UTIL_OPTIONAL = "?";
+    // Char Boolean are not primitive type, but non null equivalent jvm primitive types.
+    private static final Set<String> KOTLIN_PRIMITIVE_TYPES = new HashSet<>(asList("Byte", "Short", "Int", "Long", "Float", "Double", "Char", "Boolean"));
+
+    private final ValueMapper valueMapper;
+
+    public KotlinGraphQLTypeMapper(ValueMapper valueMapper) {
+        this.valueMapper = valueMapper;
+    }
+
+    public static boolean isKotlinPrimitive(String scalaType) {
+        return KOTLIN_PRIMITIVE_TYPES.contains(scalaType);
+    }
+
+    @Override
+    public String wrapIntoList(MappingContext mappingContext, String type, boolean mandatory) {
+        if (!mandatory && !type.endsWith("?")) {
+            return getGenericsString(mappingContext, KOTLIN_UTIL_LIST, type + "?");
+        }
+        return getGenericsString(mappingContext, KOTLIN_UTIL_LIST, type);
+    }
+
+    @Override
+    public String wrapSuperTypeIntoList(MappingContext mappingContext, String type, boolean mandatory) {
+        if (!mandatory && !type.endsWith("?")) {
+            return getGenericsString(mappingContext, KOTLIN_UTIL_LIST, "out " + type + "?");
+        }
+        return getGenericsString(mappingContext, KOTLIN_UTIL_LIST, "out " + type);
+    }
+
+    @Override
+    public String wrapApiReturnTypeIfRequired(MappingContext mappingContext,
+                                              NamedDefinition namedDefinition,
+                                              String parentTypeName) {
+        String computedTypeName = namedDefinition.getJavaName();
+        if (parentTypeName.equalsIgnoreCase(GraphQLOperation.SUBSCRIPTION.name()) &&
+                Utils.isNotBlank(mappingContext.getSubscriptionReturnType())) {
+            // in case it is subscription and subscriptionReturnType is set
+            return getGenericsString(mappingContext, mappingContext.getSubscriptionReturnType(), computedTypeName);
+        }
+        // Consider only Java-kotlin OR Java-Scala cross language calls
+        if (Boolean.TRUE.equals(mappingContext.getUseOptionalForNullableReturnTypes()) && !namedDefinition.isMandatory()) {
+            if (!computedTypeName.startsWith(KOTLIN_UTIL_LIST) && !computedTypeName.startsWith(JAVA_UTIL_LIST)) {
```

Comment:
Why this special case? Lists can be nullable or not. If they are, they should be marked as such.

Also, we are definitely not wrapping in `Optional` here. I think we should just ignore this option for Kotlin, unless it actually means using `java.util.Optional` *instead* of regular nullable types.

Reply:
Options in the list are very common. I feel that this possibility needs to be preserved. As for whether to use it or not, it should be decided by the user's schema

Reply:
on the other hand, such as
```
type A {
    a:[Int]
    b:[Int!]
    c:[Int!]!
    d:[Int]!
}
```

All of them should be allowed because of the semantics of graphql. Scala, because there is no default non null setting, only handles primitive types, but kotlin can fully resolve the semantics of graphql for all types.
Otherwise, we will lose some semantics of graphql artificially.

Reply:
So it looks like the List case is handled later by `getTypeConsideringPrimitive()`, which also deals with appending `?` (but only if not already there).
I think there is definitely no need to handle this nullability here, and we should just let `getTypeConsideringPrimitive` deal with it.
The point of having a piece of code here (I think) is only if we want to given a different meaning to `useOptionalForNullableReturnTypes`, which I mentioned in my point 2.

Reply:
Sorry, the comment is directly modified when copying, which conflicts with the code later.

The optional function of Java is limited. If you consider that option and nullable are different, there may be```Option< type?>```, which I think is ambiguous. So I wonder if it is possible to temporarily not support ```useOptionalForNullableReturnTypes ``` ？(Just like Scala's option, the implementation in the template is very tedious due to the Java code in the template). At the same time, I will remove this special logic, and make sure that they have been handled by ```getTypeConsideringPrimitive ```.

Reply:
> The optional function of Java is limited. If you consider that option and nullable are different, there may be `Option<type?>`, which I think is ambiguous.

There is no reason for `Optional<Type?>` to be generated. We control the generation. The only input we can have is a nullable or non-nullable graphql type. My point is that the `useOptionalForNullableReturnTypes` in the context of Kotlin should either be ignored (like you said) or simply make us generate `java.util.Optional<T>` *instead of* `T?`, but we would of course never generate `Optional<T?>`.

I agree we can simply ignore the option for now, but since it exists it doesn't cost much to just support it in the near future.

Reply:
I have completed the compatibility of four situations.```wrapIntoList``` and```wrapSuperTypeIntoList``` can handle most cases, but with some exceptions, I implemented them in ```getTypeConsideringPrimitive```
```kotlin
@javax.annotation.Generated(
    value = ["com.kobylynskyi.graphql.codegen.GraphQLCodegen"],
    date = "2020-12-31T23:59:59-0500"
)
interface QueryResolver {

    @Throws(Exception::class)
    fun events(): List<Event>?

    @Throws(Exception::class)
    fun event(): Event?

    @Throws(Exception::class)
    fun null1Query(): Int?

    @Throws(Exception::class)
    fun null2Query(): List<Int?>?

    @Throws(Exception::class)
    fun null3Query(): List<Int>?

    @Throws(Exception::class)
    fun null4Query(): List<Int>

    @Throws(Exception::class)
    fun null5Query(): List<Int?>

}
```

Fortunately, it doesn't contain ```Optional```, otherwise I feel suffocated